### ISSUE TITLE
System.IO.Compression handles extended characters incorrectly

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/runtime#43231]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/1190338/systemiocompression-handles-extended-characters-in.html)._

---
[severity:It's more difficult to complete my work]
create a text file name "tämä.txt"
send to compressed folder
Extracting this file it is read by library as "t„m„.txt"

Windows explorer expands the zip fine, as other compression software's

```
using System.IO;
using System.IO.Compression;

namespace ZIPtest
{
    class Program
    {
        static void Main(string[] args)
        {
            string zipFile = @"C:\temp\tämä.zip";

using (ZipArchive paketti = ZipFile.OpenRead(zipFile))
            {
                foreach (ZipArchiveEntry entry in paketti. Entries)
                {
                    string nimi = Path.Combine(@"c:\temp", entry. Name);
                    entry. ExtractToFile(nimi);
                }

}
        }
    }
}
```

---
#### Feedback Bot on 9/20/2020, 11:00 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
#### Tarek Mahmoud Sayed [MSFT] solved on 9/30/2020, 01:44 PM, 0 votes: 

<p>Thanks for sending the issue. I have tried it the same thing and I was not able to reproduce the issue. My guess here is the problem is not really the <a target='_blank' href="http://ZipArchiveEntry.Name">ZipArchiveEntry.Name</a> content but it could be the Visual Studio locals Window displaying the string differently and this can be depending on the configuration of your machine. Usually I am seeing this maybe depending on the default codepage on your system.</p>
<p>Here is the code I tried:</p>
<pre><code>        using (ZipArchive paketti = ZipFile.OpenRead(@&quot;F:\temp\release.zip&quot;))
        {
            foreach (ZipArchiveEntry entry in paketti.Entries)
            {
                Console.WriteLine($&quot;{entry.Name} ... {DumpString(entry.Name)}&quot;);
            }
        }
</code></pre>
<p>And this printed the output:</p>
<pre><code>  tämä.txt ... \u0074\u00e4\u006d\u00e4\u002e\u0074\u0078\u0074
</code></pre>
<p>which is correct.</p>
<p>I suggest you can try the same either manually printing the characters ordinal values and check your system configuration like default locale and codepage in the system.</p>

####  on 10/8/2020, 05:05 AM: 

(private comment, text removed)
#### Tarek Mahmoud Sayed [MSFT] on 10/8/2020, 10:10 AM: 

<p>Thanks for your reply. Unfortunately I cannot reach the files that you have attached. Could you please try to upload them again and let me know. <br><br>Also, please attach the code that you used to create the zip file again to see how did you include the file tämä.txt too.</p>

####  on 10/9/2020, 02:29 AM: 

(private comment, text removed)
#### Tarek Mahmoud Sayed [MSFT] on 10/9/2020, 11:01 AM: 

<p>Thanks for the details. I'll take a look.</p>

**Comment by tarekgh** [comment]
I have looked at this issue and I am seeing the Zip archive file name encoding is not handled correctly. When creating a new archive file there will be a generic flag telling if the archive is encoded using UTF-8 or not. if this flag is off, means the archive is not encoded using UTF-8, we don't handle the right encoding at that time. the following comment has more details:

https://github.com/dotnet/runtime/blob/b75f8d9fd1b3bcac5d82469bf39f1aea2c3a1652/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs#L356

Usually when UTF-8 encoding is not used, we should consider using the 437 encoding instead. Or try to investigate more the details as described in https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT in `Appendix D`. 

This issue is kind of obvious to the users because when using Windows shell to create a new Archive, it doesn't create the archive using UTF-8 and I am seeing it is using encoding 437.

**Comment by jyrijh** [comment]
Thank you. Now I know what is happening and this will be fixed.

But in the mean time I could do this, and force codepage 437 to get correct value from `_storedEntryNameBytes`.
`entry.Name` stays incorrect.

I have to test if this works correctly in all cases

```c#
string zipFile = @"C:\temp\tämä.zip";            
var ibm437 = Encoding.GetEncoding("IBM437");

using (ZipArchive paketti = ZipFile.OpenRead(zipFile))
{
    var entryEncoding = paketti.GetType().
        GetProperties(BindingFlags.NonPublic | BindingFlags.Instance).
        Single(pi => pi.Name == "EntryNameEncoding");

    entryEncoding.SetValue(paketti, ibm437);

    foreach (ZipArchiveEntry entry in paketti.Entries)
    {
        var name = entry.GetType().GetFields(BindingFlags.NonPublic | BindingFlags.Instance).Single(pi => pi.Name == "_storedEntryNameBytes");
        var nameValue = name.GetValue(entry);
        var result = ibm437.GetString((byte[])nameValue);

        Console.WriteLine($"{entry.Name} ... {DumpString(entry.Name)}");
        Console.WriteLine($"{result} ... {DumpString(result)}");
    }
}
```

**Comment by jyrijh** [comment]
```c#
ZipFile.Open(String, ZipArchiveMode, Encoding)
```

Now I just need to figure out what encoding has been used

**Comment by jyrijh** [comment]
With this I came up. Minimal detection for encoding and after quick testing it works.

```c#
using System;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;

namespace ZIPtest
{
    class Program
    {
        [Flags]
        public enum GeneralPurpose : ushort
        {
            None = 0,
            bit00 = 1 << 0,
            bit01 = 1 << 1,
            bit02 = 1 << 2,
            bit03 = 1 << 3,
            bit04 = 1 << 4,
            bit05 = 1 << 5,
            bit06 = 1 << 6,
            bit07 = 1 << 7,
            bit08 = 1 << 8,
            bit09 = 1 << 9,
            bit10 = 1 << 10,
            languageEncoding = 1 << 11,
            bit12 = 1 << 12,
            bit13 = 1 << 13,
            bit14 = 1 << 14,
            bit15 = 1 << 15
        }

        [StructLayout(LayoutKind.Sequential, Pack = 1)]
        struct Header
        {
            [MarshalAs(UnmanagedType.ByValArray, SizeConst = 4)]
            public byte[] signature;
            public ushort version;
            public GeneralPurpose generalPurposeBit;
        }

        static void Main()
        {
            string filename = @"C:\temp\tämä.zip";
            var encoding = DetectEncoding(filename);
            CheckNames(filename, encoding);

            filename = @"C:\temp\漢字.zip";
            encoding = DetectEncoding(filename);
            CheckNames(filename, encoding);

            void CheckNames(string zipName, Encoding zipEncoding)
            {
                using (ZipArchive paketti = ZipFile.Open(zipName, ZipArchiveMode.Read, zipEncoding))
                {
                    foreach (ZipArchiveEntry entry in paketti.Entries)
                    {
                        Console.WriteLine($"{entry.Name} ... {DumpString(entry.Name)}");
                    }
                }
            }

            string DumpString(string name) => string.Join(", ", name.ToArray().Select(x => $"{(int)x:X4}"));
        }


        static Encoding DetectEncoding(string file)
        {
            var data = new byte[8];
            using (FileStream sr = new FileStream(file, FileMode.Open))
                sr.Read(data, 0, 8);

            GCHandle gch = GCHandle.Alloc(data, GCHandleType.Pinned);
            Header header = (Header)Marshal.PtrToStructure(gch.AddrOfPinnedObject(), typeof(Header));
            gch.Free();

            if (header.generalPurposeBit.HasFlag(GeneralPurpose.languageEncoding))
                return Encoding.UTF8;
            else
                return Encoding.GetEncoding("IBM437");
        }
    }
}
```

**Comment by jyrijh** [comment]
And no it didn't work as intended.

if zip file has these in it, they will appear in this order, and my code will get encoding from first entry.
```
aa 漢字.txt
äÄöÖåÅüÜéÉ.txt
漢字.txt
```
And then the second filename is actually 437 encoded. 
IO.Compression will handle everything with same encoding when it should handle it per entry.

**Comment by tarekgh** [comment]
`ZipFile.Open(String, ZipArchiveMode, Encoding)` can be used to choose the specific encoding and should help if you know the encoding.

For https://github.com/dotnet/runtime/issues/43231#issuecomment-707643592, this may work in some cases but not in general cases as you indicated later. if you look at pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT it says 

```
D.3 Applications MAY choose to supplement this file name storage through the use 
of the 0x0008 Extra Field.  Storage for this optional field is currently 
undefined, however it will be used to allow storing extended information 
on source or target encoding that MAY further assist applications with file 
name, or file content encoding tasks.  Please contact PKWARE with any
requirements on how this field SHOULD be used.
```

**Comment by jyrijh** [comment]
After giving this some thought, exposing the header for entries would be suitable solution for this. Then it would be up to programmer to check those values if they are interested in them.

**Comment by jyrijh** [comment]
This is exactly why the language encoding bit should be exposed. if it is other that UTF-8 then runtime has no way of knowing what the codepage should be. Me as developer can "guess" what codepage to use by knowing where the file is coming from.

**Comment by cremor** [comment]
> if it is other that UTF-8 then runtime has no way of knowing what the codepage should be

How are standalone archiving tools like 7-Zip are handling that case then? 7-Zip can extract all archives that I tested with correct encodings. If 7-Zip can do it then the .NET runtime can do it for sure too?

**Comment by jyrijh** [comment]
If you create an archive that has any other encoding that UTF-8 or IBM437 it doesn't.

7-zip will extract this ")=.txt" as "]~Kúºú"

So I  need to know where the file came from and have knowledge what codepage they used.
Usually it should be UTF-8 or IBM437, but even then I need to know then per entry what the language encoding bit is

```c#
var zipFile = @"c:\temp\test.zip";
var name = ")=.txt";
var encoding = Encoding.GetEncoding("IBM297");

using (FileStream stream = new FileStream(zipFile, FileMode.OpenOrCreate))
{
    using (ZipArchive archive = new ZipArchive(stream, ZipArchiveMode.Update, false, encoding))
    {
        var entry = archive.CreateEntry(name);
        using (StreamWriter writer = new StreamWriter(entry.Open()))
        {
            writer.WriteLine("text");
        }
    }
}
```

**Comment by cremor** [comment]
Yeah ok, but I meant archives that were created with "standard" tools. .NET should be able to handle those automatically.

**Comment by cremor** [comment]
I just found this comment in the .NET source code: https://source.dot.net/#System.IO.Compression/System/IO/Compression/ZipArchive.cs,356

What's weird about that is that it seems to be wrong. Specifically the part where it says

> For instance, the Windows Shell Zip tool takes "something else" to mean "the local system codepage". We default to the same behaviour, [...]

In my experience this is not the case, see https://github.com/dotnet/runtime/issues/43231#issuecomment-1008834276. The Windows Shell Zip tool seems to use "IBM850" (**not** "IBM437"!) encoding (on Windows 10). And because .NET does **not** default to that (but instead defaults to ["the current system default code page"](https://docs.microsoft.com/en-us/dotnet/api/system.io.compression.ziparchive.-ctor?view=netframework-4.8#system-io-compression-ziparchive-ctor(system-io-stream-system-io-compression-ziparchivemode-system-boolean-system-text-encoding))), those Windows zip files can not be correctly extracted by .NET.

Maybe .NET just understands "the current system default code page" different than Windows. As written, Windows seems to use "IBM850" encoding, which is available in .NET as `CultureInfo.CurrentCulture.TextInfo.OEMCodePage`. But .NET seems to think that the default is "Windows-1252" (available in `Encoding.Default`).

Anyway, even if it would work by default for zip files created on Windows, we would still have problems with zip files created on macOS.

**Comment by tarekgh** [comment]
> Maybe .NET just understands "the current system default code page" different than Windows. As written, Windows seems to use "IBM850" encoding, which is available in .NET as CultureInfo.CurrentCulture.TextInfo.OEMCodePage. But .NET seems to think that the default is "Windows-1252" (available in Encoding.Default).

If you are running with full framework, default code page would be picked up from Windows API `GetACP`. If running with .NET or .NET Core, the default codepage would be UTF-8. to get the default system codepage in .NET, need to register the codepages encoding as described in https://docs.microsoft.com/en-us/dotnet/api/system.text.codepagesencodingprovider?view=net-6.0#remarks and then call `Encoding.GetEncoding(0)`.

**Comment by cremor** [comment]
I tested on .NET Framework 4.8.

But as far as I understand it .NET Core/5/6 will have the same problem, right? Because neither uses the same ZIP entry name encoding as Windows does.

**Comment by tarekgh** [comment]
> But as far as I understand it .NET Core/5/6 will have the same problem, right? 

Yes. I was just explaining how Encoding picks the default.

**Comment by caverna** [comment]
> I tested on .NET Framework 4.8.
> 
> But as far as I understand it .NET Core/5/6 will have the same problem, right? Because neither uses the same ZIP entry name encoding by default that Windows uses.

yep! I reached this page having the same issue and it fails for .NET Core 7

**Comment by Stallind** [comment]
Commeting here since I found the thread googling about the same issue.

For .net framework 4.8 it works for me using Encoding.GetEncoding(System.Globalization.CultureInfo.CurrentCulture.TextInfo.OEMCodePage).

For .net core I used a somewhat ugly hack where I replace the "illegal characters". In this case swedish åäö.

```c#
var replacements = new Dictionary<string,string> { 
    { "\u0086", "å" },
    { "\u008f", "Å"},
    { "\u0084", "ä"},
    { "\u008e", "Ä"},
    { "\u0094", "ö"},
    { "\u0099", "Ö"} 
};
using (ZipArchive source = System.IO.Compression.ZipFile.Open(@"C:\New Folder\testing.zip", ZipArchiveMode.Read, Encoding.GetEncoding("ISO-8859-1")))
{
    foreach (ZipArchiveEntry entry in source.Entries)
    {
        var name = entry.Name;
        foreach (var replacement in replacements)
        {
            name = name.Replace(replacement.Key, replacement.Value);
        }
        Console.WriteLine(name);
    }
}
```

I can't say this is the way to go and that I recommend it, but it solved the issue for me.

**Comment by alinpahontu2912** [comment]
Hey @jyrijh  @Stallind @tarekgh @cremor, doesn't [this](https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/9.0/ziparchiveentry-encoding) mean that the problem has been solved already?

**Comment by cremor** [comment]
Maybe, I haven't tested this with .NET 9 yet. My app where I found this problem uses .NET Framework and will probably never be ported.

**Comment by karelz** [comment]
The bar for changing .NET Framework is high. There is significant risk with every change that it will break something. Therefore, we fix only high-impactful / security issues in .NET Framework. For latest innovation, performance and cool new feature, .NET (Core) is the right place to be.

**Comment by jyrijh** [comment]
No, the zip entries don't expose the general purpose bit flag, there is no way the determine entry encoding in .NET.
And ZipArchive doesn't offer automatic detection.

So if i try to open zip files without encoding, it defaults to uft-8 and entries that are in different encoding will come out incorrectly.

<img width="1108" height="551" alt="Image" src="https://github.com/user-attachments/assets/9c8d08ce-b83c-49db-88f2-7a9232f66d4b" />

So you have to write you own encoding detection. What i made only checks the first entry from central directory header and based on that I assume that all other entries  filename encoding will be the same

**Comment by jyrijh** [comment]
So you could also use reflection and get the `general purpose bit flag` from entry and based on that you can make decisions should you use the name as is or convert the bytes to another encoding.

Code below will create this output
```text
äÄöÖåÅüÜéÉ - Copy.txt ... äÄöÖåÅüÜéÉ - Copy.txt UnicodeFileNameAndComment
??????????.txt ... äÄöÖåÅüÜéÉ.txt 0
```

```csharp
using System.IO.Compression;
using System.Reflection;
using System.Text;

Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
string zipFile = @"C:\temp\mixed.zip";
var ibm437 = Encoding.GetEncoding("IBM437");

using ZipArchive paketti = ZipFile.Open(zipFile, ZipArchiveMode.Read);
var entryEncoding = paketti.GetType();

foreach (ZipArchiveEntry entry in paketti.Entries)
{
    var name = entry.GetType().GetFields(BindingFlags.NonPublic | BindingFlags.Instance).Single(pi => pi.Name == "_storedEntryNameBytes");

    BitFlagValues generalPurposeBitFlag = (BitFlagValues)entry.GetType()
        .GetFields(BindingFlags.NonPublic | BindingFlags.Instance)
        .Single(pi => pi.Name == "_generalPurposeBitFlag")
        .GetValue(entry)!;

    string result = null!;
    if ((generalPurposeBitFlag & BitFlagValues.UnicodeFileNameAndComment) != 0)
        result = entry.Name;
    else
        result = ibm437.GetString((byte[])name.GetValue(entry)!);

    Console.WriteLine($"{entry.Name} ... {result} {generalPurposeBitFlag}");
}

[Flags]
internal enum BitFlagValues : ushort
{
    IsEncrypted = 0x1,
    DataDescriptor = 0x8,
    UnicodeFileNameAndComment = 0x800
}
```

mixed.zip was created with this.

```csharp
using System.IO.Compression;
using System.Text;

Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);

{
    using ZipArchive paketti = ZipFile.Open(@"C:\temp\mixed.zip", ZipArchiveMode.Create, Encoding.UTF8);
    var entry = paketti.CreateEntryFromFile(@"C:\temp\äÄöÖåÅüÜéÉ - Copy.txt", "äÄöÖåÅüÜéÉ - Copy.txt");
}

{
    using ZipArchive paketti = ZipFile.Open(@"C:\temp\mixed.zip", ZipArchiveMode.Update, Encoding.GetEncoding("IBM437"));
    var entry = paketti.CreateEntryFromFile(@"C:\temp\äÄöÖåÅüÜéÉ.txt", "äÄöÖåÅüÜéÉ.txt");
}
```

**Comment by jyrijh** [comment]
And this works only in .NET. I tested with Framework version 4.7 -> 4.8.1 and they always return the encoding you opened the zip file for `_generalPurposeBitFlag`. So in there you have to assume that the whole file has the same encoding as first entry.

**Comment by alinpahontu2912** [comment]
Hey @jyrijh, thanks for the comments. As mentioned above, touching .NET Framework is unlikely to happen. The new behavior of the ZipArchive class corrects a bug that was present in .NET 7 and .NET 8. Previously, the library would ignore the UTF-8 bit flag leading to corrupted filenames with special characters.
Starting in .NET 9, the library now correctly checks each entry's header for the UTF-8 bit flag. If the flag is set, it uses UTF-8 to decode the filename. The user-supplied encoding is only used if the flag is not present. This change makes the ZipArchive library compliant with the ZIP specification and eliminates the need for reflection-based workarounds to handle mixed encodings.

**Comment by jyrijh** [comment]
@alinpahontu2912 but it still doesn't help if you have a zip file that was created like

```csharp
using ZipArchive paketti = ZipFile.Open(@"C:\temp\IBM437.zip", ZipArchiveMode.Create, Encoding.GetEncoding("IBM437"));
var entry = paketti.CreateEntryFromFile(@"C:\temp\äÄöÖåÅüÜéÉ.txt", "äÄöÖåÅüÜéÉ.txt");
```

And then you open it without providing encoding, then it is now opened always with utf-8, which will produce wrong output.
And if you have a mixed zip, then those that were with utf-8 are correct, but rest depends.

For me the good solution would be, if the `generalPurposeBitFlag` and `storedEntryNameBytes` were properties for `ZipArchiveEntry` so I could check per entry, and act accordingly, so I don't need to use reflection, or read the bytes from file

**Comment by jyrijh** [comment]
For now i created for myself this extension class

```csharp
public static class ZipExtensions
{
    [Flags]
    internal enum BitFlagValues : ushort
    {
        IsEncrypted = 0x1,
        DataDescriptor = 0x8,
        UnicodeFileNameAndComment = 0x800
    }

    public static bool IsUnicode(this ZipArchiveEntry entry)
    {
        BitFlagValues generalPurposeBitFlag = (BitFlagValues)entry.GetType()
            .GetFields(BindingFlags.NonPublic | BindingFlags.Instance)
            .Single(pi => pi.Name == "_generalPurposeBitFlag")
            .GetValue(entry)!;

        return (generalPurposeBitFlag & BitFlagValues.UnicodeFileNameAndComment) != 0;
    }

    public static string GetDecodedName(this ZipArchiveEntry entry, Encoding encoding)
    {
        // _storedEntryNameBytes is needed if we need to decode the name from another encoding
        var name = entry.GetType().GetFields(BindingFlags.NonPublic | BindingFlags.Instance).Single(pi => pi.Name == "_storedEntryNameBytes");
        return encoding.GetString((byte[])name.GetValue(entry)!);
    }
}
```

**Comment by rzikm** [comment]
> [@alinpahontu2912](https://github.com/alinpahontu2912) but it still doesn't help if you have a zip file that was created like
> 
> using ZipArchive paketti = ZipFile.Open(@"C:\temp\IBM437.zip", ZipArchiveMode.Create, Encoding.GetEncoding("IBM437"));
> var entry = paketti.CreateEntryFromFile(@"C:\temp\äÄöÖåÅüÜéÉ.txt", "äÄöÖåÅüÜéÉ.txt");
> And then you open it without providing encoding, then it is now opened always with utf-8, which will produce wrong output. And if you have a mixed zip, then those that were with utf-8 are correct, but rest depends.

Do I understand correctly that you believe .NET should use IBM437 encoding by default (specifically, unless UnicodeFileNameAndComment bit is set and user did not explicitly provide other encoding)?

I think that would bring us more in line with the specification. @dotnet/area-system-io-compression do we see any issue with such a change? I think we can keep a preference to UTF8 when writing non-ascii names (if explicit encoding wasn't specified), but for reading, I think we should default to IBM437 for non-UTF8 entries.

**Comment by jyrijh** [comment]
@rzikm I'm personally ok with UTF8, but what I want is a way to know if it was stored with UFT8 or not. And if it wasn't then I can make decisions based on that information.

As if it wasn't then I have either agreed with creator that they use specific encoding, and I can use the same for decoding the filename.

**Comment by alinpahontu2912** [comment]
Hey @cremor @vsfeedback @jyrijh , after some internal discussions, this is our thinking:
ZIP archives do not record the original code page used for entry names, they only indicate whether each name is encoded in UTF-8 (bit 11) or in an unspecified encoding. Because the specific non UTF-8 encoding is not preserved, consumers must provide it explicitly when reading such archives.
If you are working with ZIP files that use non UTF-8 or other encodings, the overloads that accept an Encoding parameter should resolve any issues related to entry name decoding.
```c#
// With file: 
    using var z1 = System.IO.Compression.ZipFile.Open(
    path, ZipArchiveMode.Read, Encoding.GetEncoding("Latin1"));
// With Stream: 
    using var z2 = new System.IO.Compression.ZipArchive(
    stream, ZipArchiveMode.Read, leaveOpen: false, entryNameEncoding: Encoding.GetEncoding(1250));
```
Extra clarifications:

- When the UTF-8 flag is set, UTF-8 is always used, regardless of the encoding specified. This behavior is defined by the ZIP specification.
- When the UTF-8 flag is not set, the encoding you provide will be used.

This is already being checked internally.

In summary, if a ZIP archive uses a non-standard encoding, specifying the correct Encoding when opening the archive should solve the issue.

References:
- 
- .NET behavior [documentation](https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/9.0/ziparchiveentry-encoding)
- Overloads supporting Encoding parameter: [ZipFile.Open](https://learn.microsoft.com/en-us/dotnet/api/system.io.compression.zipfile.open?view=net-9.0), [Ziparchive constructor](https://learn.microsoft.com/en-us/dotnet/api/system.io.compression.ziparchive.-ctor?view=net-9.0)
